compiler/noirc_evaluator/src/ssa/opt/runtime_separation.rs

```diff
@@ -0,0 +1,350 @@
+    /// After this step, all functions with runtime `Acir` will be converted to Acir and
+    /// the functions with runtime `Brillig` will be converted to Brillig.
+    /// It does so by cloning all ACIR functions called from a Brillig context
+    /// and changing the runtime of the cloned functions to Brillig.
+    /// This pass needs to run after functions as values have been resolved (defunctionalization).
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn separate_runtime(mut self) -> Self {
+        RuntimeSeparatorContext::separate_runtime(&mut self);
+
+        self
+    }
+}
+
+#[derive(Debug, Default)]
+struct RuntimeSeparatorContext {
+    // Original functions to clone to brillig
+    acir_functions_called_from_brillig: BTreeSet<FunctionId>,
+    // Tracks the original => cloned version
+    mapped_functions: HashMap<FunctionId, FunctionId>,
+}
+
+impl RuntimeSeparatorContext {
+    pub(crate) fn separate_runtime(ssa: &mut Ssa) {
+        let mut runtime_separator = RuntimeSeparatorContext::default();
+
+        // We first collect all the acir functions called from a brillig context by exploring the SSA recursively
+        let mut processed_functions = HashSet::default();
+        runtime_separator.collect_acir_functions_called_from_brillig(
+            ssa,
+            ssa.main_id,
+            false,
+            &mut processed_functions,
+        );
+
+        // Now we clone the relevant acir functions and change their runtime to brillig
+        runtime_separator.convert_acir_functions_called_from_brillig_to_brillig(ssa);
+
+        // Now we update any calls within a brillig context to the mapped functions
+        runtime_separator.replace_calls_to_mapped_functions(ssa);
+
+        // Some functions might be unreachable now (for example an acir function only called from brillig)
+        prune_unreachable_functions(ssa);
+    }
+
+    fn collect_acir_functions_called_from_brillig(
+        &mut self,
+        ssa: &Ssa,
+        current_func_id: FunctionId,
+        mut within_brillig: bool,
+        processed_functions: &mut HashSet<(/* within_brillig */ bool, FunctionId)>,
+    ) {
+        // Processed functions needs the within brillig flag, since it is possible to call the same function from both brillig and acir
+        if processed_functions.contains(&(within_brillig, current_func_id)) {
+            return;
+        }
+        processed_functions.insert((within_brillig, current_func_id));
+
+        let func = ssa.functions.get(&current_func_id).expect("Function should exist in SSA");
+        if func.runtime() == RuntimeType::Brillig {
+            within_brillig = true;
+        }
+
+        let called_functions = called_functions(func);
+
+        if within_brillig {
+            for called_func_id in called_functions.iter() {
+                let called_func =
+                    ssa.functions.get(called_func_id).expect("Function should exist in SSA");
```

Comment:
```suggestion
                let called_func = &ssa.functions[called_func_id];
```
Nitpick: we do this in a few places but I don't think the expect message adds much here